### PR TITLE
fix(parsers): preserve duration-authored keyframe timing

### PR DESCRIPTION
**Pull request by miguel-heygen**
## What

Preserve duration-authored timing when parsing and editing GSAP object-array keyframes.

This is Family A, PR 4 of 4. Prerequisite: `codex/studio-timeline-a-ease-editor-v2`. It is the Family A tip.

Review context: #2582 and the post-certification review fixes for #2387.

## Why

Object-array keyframes use per-step durations as their timing source. Treating them as evenly distributed percentages can create, update, or remove a keyframe at the wrong authored position.

## How

A single timing helper now owns duration-to-percentage conversion and index resolution for both the recast and Acorn parsers/writers.

Review hardening preserves historical tenth-percent precision, rejects zero/negative/non-finite/expression/mixed duration sequences without rewriting source, refuses lossy conversion when outer duration conflicts with the per-step total, preserves the historical nearest-entry update behavior, and retains tolerance-based deletion so a fuzzy delete cannot remove the wrong keyframe.

The scope is timing fidelity for the supported keyframe/ease contracts; it does not claim arbitrary multi-segment custom paths or per-property easing.

## Test plan

- [x] Unit tests added/updated — 528 parser tests passed (524 active, 4 intentional skips) across both parser/writer implementations
- [x] Boundary coverage added — precision, malformed duration, unresolved expression, outer mismatch, nearest update, and safe deletion
- [x] Manual testing performed — duration-authored add, update, remove, and round-trip behavior is covered by the certified integrated QA matrix
- [x] Full workspace production build passed on latest `main`
- [ ] Documentation updated (not applicable; internal parser behavior)

### CODE REVIEW

**Comment by miguel-heygen** [comment]
* **#2675** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/2675?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/2675?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2674** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/2674?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2673** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/2673?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2672** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/2672?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by vanceingalls — DISMISSED** [review]
## Verdict: APPROVE (with one precision-regression note to confirm)

**Rule applied:** RIGHT-direction non-blocker findings → APPROVE (per COMMENTED-vs-APPROVE gate). Duration-authored parsing/writing was the correct fix; the extraction is clean.

**Multi-reviewer inventory:** parallel review with Rames. No other review at submission time.

### What holds up

- **`gsapObjectArrayTiming.ts` extraction is the right move.** Duration→percentage conversion was previously duplicated across `parseObjectArrayKeyframes`, `convertArrayKeyframesToObjectNode`, `removeKeyframeFromScript`, and `updateKeyframeInScript`. Centralizing means the parser, all writers, and the index resolver agree by construction — no drift possible when authored duration is what times the keyframe.
- **`totalDuration` return field** — surfaces the synthesized outer-duration signal to `convertArrayKeyframesToObjectNode:2115-2117`, which now writes it back only when the source did not author one. Correct semantics for "duration authored" vs "duration synthesized".
- **`findObjectArrayKeyframeIndex` tie-break** — smallest distance wins, ties resolved by "first seen". Deterministic. Matches the closest-match tolerance model the existing `PCT_TOLERANCE` callers used.
- **`convertArrayKeyframesToObjectNode` per-element duration strip (`gsapParser.ts:2115-2119`)** — pulls `duration` out of each keyframe object before emitting the `{"25%": {...}}` form. Prevents the runtime from receiving a stale per-step `duration` in an object-form keyframe (where GSAP would ignore it) and preserves the authored total via the outer `duration` write. Well done.
- **Test coverage matrix** — `gsapWriter.parity.test.ts` covers duration-authored remove, update, move, and resize on both writers with reparse-equivalence assertions. 434 tests passing with 4 documented skips.

### One regression to confirm

**Precision loss in the object-form percentages emitted by `convertArrayKeyframesToObjectNode`.** Before this PR (`gsapParser.ts:2113`):

```ts
const pct = n > 1 ? Math.round((i / (n - 1)) * 1000) / 10 : 0;
```

That was **1-decimal precision** (`33.3%`, `66.7%`). After the extraction, the unified helper `getObjectArrayKeyframeTiming` uses `Math.round((cumulative / totalDuration) * 100)` — **integer precision only** (`33%`, `67%`).

The parity tests reparse-round-trip so they pass either way. But downstream consumers that store keyframe percentages as authored strings (e.g. `"33.3%"`) will see a shape change: a 3-entry array-form tween that previously converted to `{"0%":…, "33.3%":…, "66.7%":…, "100%":…}` now converts to `{"0%":…, "33%":…, "67%":…, "100%":…}`.

For a duration-authored source this is fine — the whole point of this PR is that `33.3%` was the wrong invented percentage anyway. For an evenly-distributed source (no per-step duration), you're intentionally accepting the precision drop as the unification cost. Please confirm:

1. No fixture / golden compares against the pre-PR `"33.3%"` string form.
2. Downstream Studio UI is OK with integer-only percentages on evenly-distributed converts (drag targets, snap resolution, keyframe list rendering).

If the answer to both is yes, ship it. Otherwise a `precision: "integer" | "tenth"` option on `getObjectArrayKeyframeTiming` would preserve the pre-PR behavior for the even-distribution branch.

### Nit

**`gsapParser.ts:2103-2106` — `objectExpressionToRecord(element, scope)` on every keyframe element inside `convertArrayKeyframesToObjectNode`.** Runs on the FULL record just to fish out `duration`. If keyframes are large (many properties), this walks each element twice — once to record, once to filter properties on the emit side. Micro-optimization; probably invisible in practice. Skip if you like.

### Parser-schema alignment with #2673

- `mergePercentageKeyframes` (#2673) and the object-array timing helper (#2675) do not touch overlapping code paths — no merge conflict shape.
- Both PRs share `objectExpressionToRecord` as the vars-node reader. Contract unchanged, safe.
- The `moveKeyframeInScript` no-op-on-collision change from #2673 is preserved through #2675's normalize-then-move path in the duration-authored test (`gsapWriter.parity.test.ts:1284-1291`).

— Via

File: packages/parsers/src/gsapObjectArrayTiming.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        return Math.round((cumulative / totalDuration) * 100);
+      }),
+      totalDuration,
+    };
+  }
+
+  const lastIndex = durations.length - 1;
+  return {
+    percentages: durations.map((_, index) =>
+      lastIndex > 0 ? Math.round((index / lastIndex) * 100) : 0,
```
Comment (james-russo-rames-d-jusso) [inline packages/parsers/src/gsapObjectArrayTiming.ts]: 🟠 Precision regression on even-distributed arrays — interior keyframes shift by ~0.33% at runtime after any add/move/resize/removeAll that triggers conversion.

```ts
return {
  percentages: durations.map((_, index) =>
    lastIndex > 0 ? Math.round((index / lastIndex) * 100) : 0,
  ),
};
```

The base `convertArrayKeyframesToObjectNode` and base `convertArrayKeyframesToObject` (both replaced call sites) emitted one-decimal precision via `Math.round((i/(n-1))*1000)/10`. This helper rounds to integer percentages.

Concrete failure: a 4-element even array `[a, b, c, d]` used to convert to `"0%", "33.3%", "66.7%", "100%"`. Now converts to `"0%", "33%", "67%", "100%"`. GSAP interprets those keys literally, so interior keyframes shift by ~0.33% at runtime.

**Why it matters**: the PR title is *"preserve duration-authored keyframe timing"* — but the NO-duration path is where the actual behavior changed. Any script whose author didn't specify per-step durations (relying on even distribution) will render slightly differently after any mutation trigger.

**Fix**: match the prior one-decimal precision:
```ts
percentages: durations.map((_, index) =>
  lastIndex > 0 ? Math.round((index / lastIndex) * 1000) / 10 : 0,
),
```

Same for the duration-authored branch at `:20-24` — should also emit one-decimal to match the historical output shape.

Also worth: add a regression test that a 4-element even-distributed array converts to `"0%,33.3%,66.7%,100%"` (not `"0%,33%,67%,100%"`).

— _[Review by Rames D Jusso](https://github.com/james-russo-rames-d-jusso)_

File: packages/parsers/src/gsapObjectArrayTiming.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Resolve GSAP object-array keyframe positions exactly once for parsers and writers.
+ * Authored per-step durations place each keyframe at its cumulative end; arrays
+ * without durations are distributed evenly.
+ */
+export function getObjectArrayKeyframeTiming(
+  durations: ReadonlyArray<number | undefined>,
+): ObjectArrayKeyframeTiming {
+  const totalDuration = durations.reduce<number>((sum, duration) => sum + (duration ?? 0), 0);
+  if (totalDuration > 0) {
```
Comment (james-russo-rames-d-jusso) [inline packages/parsers/src/gsapObjectArrayTiming.ts]: 🟡 Boundary cases silently accepted or silently mis-mapped:

1. **Zero-duration entry**: `[0, 1, 0]` yields cumulative `[0, 1, 1]` → percentages `[0, 100, 100]`. `findObjectArrayKeyframeIndex` breaks ties on strict `<`, so the *first* match wins; GSAP evaluates the *later* keyframe at that time. Update targets the wrong element.
2. **Negative duration**: `[1, -1, 5]` yields non-monotonic cumulative percentages `[20, 0, 100]`. Silently accepted. If two negatives cancel to `total = 0`, code falls through to even-distribute, masking the input.
3. **Non-numeric duration** (variable / expression): `valueNodeToRecord` returns `__raw: ...`, so `typeof duration === "number"` is false and the duration is treated as `undefined`. A script authored with `duration: totalMs * 0.5` collapses to even distribution — the very case this PR advertises fixing.
4. **Sum-of-durations ≠ authored outer `duration`**: `convertArrayKeyframesToObjectNode` only sets outer `duration` when absent. If both are present but disagree, the per-step values are dropped and the mismatched outer duration is retained — not a lossless conversion.

**Fix**: at the helper boundary, add:
```ts
if (durations.some(d => d !== undefined && (typeof d !== 'number' || !Number.isFinite(d) || d < 0))) {
  // signal or reject — don't silently coerce
}
```

And a test for each of the 4 boundary cases. Silent coercion of malformed input is exactly the class of bug that surfaces months later as "the animation is subtly wrong" with no traceable cause.

— _[Review by Rames D Jusso](https://github.com/james-russo-rames-d-jusso)_

File: packages/parsers/src/gsapWriterAcorn.ts [hunk 3]
```diff
@@ -955,13 +957,18 @@ function updateArrayKeyframeByPct(
...
-  const idx = n > 1 ? Math.round((percentage / 100) * (n - 1)) : 0;
-  const el = elements[Math.max(0, Math.min(n - 1, idx))];
+  if (elements.length === 0) return script;
+  const records = elements.map((element) => valueNodeToRecord(element, script));
+  const idx = findObjectArrayKeyframeIndex(
+    records.map((record) => (typeof record.duration === "number" ? record.duration : undefined)),
+    percentage,
+    PCT_TOLERANCE,
+  );
+  if (idx === null) return script;
```
Comment (james-russo-rames-d-jusso) [inline packages/parsers/src/gsapWriterAcorn.ts:998]: 🟡 Semantic change: `updateArrayKeyframeByPct` moved from index-round (always-match, clamped) to tolerance-based (returns `null` when nothing within `PCT_TOLERANCE=2`).

```ts
const idx = findObjectArrayKeyframeIndex(
  records.map((record) => (typeof record.duration === "number" ? record.duration : undefined)),
  percentage,
  PCT_TOLERANCE,
);
if (idx === null) return script;
```

Base behavior: `Math.round((percentage/100)*(n-1))` + `Math.max/min` clamp → always matched some index. New behavior: returns null if the nearest percentage is farther than 2% from target.

Concrete failure: a caller passing `percentage=50` against 4 even entries `[0, 33, 67, 100]` used to update index 2; now silently no-ops (nearest is 33 or 67, both 17% away).

Likely intentional (removes the fuzzy-clamp footgun) but worth calling out — any Studio drag path that emits fuzzy percentages (drag snap-to-percent) will change behavior. The parity test at `gsapWriter.parity.test.ts` doesn't cover the mid-gap case.

**Fix**: document the semantic shift in the calling comment, OR add a fallback to nearest-index-clamped when tolerance-match returns null AND `percentage ∈ [0, 100]`.

Minor: `PCT_TOLERANCE = 2` is a magic number — no named constant with rationale. If Studio ever drags with sub-1%-resolution (which the curve editor at #2674 might, given `round2`), the tolerance may need tuning.

— _[Review by Rames D Jusso](https://github.com/james-russo-rames-d-jusso)_

**Review by james-russo-rames-d-jusso — COMMENTED** [review]
_Reviewed at `503372f0125536d6c00957cab72182dac6c30387`._

Family A tip — extracts a shared `getObjectArrayKeyframeTiming` helper for GSAP object-array keyframes, and routes the recast parser + Acorn writer through it for parse/convert/remove/update. The merge-instead-of-replace change on the array-form update path is a real correctness fix (replacing would drop authored `duration` and cascade cumulative shifts to siblings). Tests are added at 75% against `[1, 2, 1]` fixtures.

**No blockers**. Two structural concerns worth folding in before merge; several boundary omissions.

### Cross-cutting theme: partial helper adoption + precision regression

The PR body says *\"a single timing helper owns duration-to-percentage conversion and index resolution\"* — but two things undermine that:

1. **Acorn parser wasn't migrated** (`gsapParserAcorn.ts:1022`). The recast parser routes through the new helper; the Acorn parser still inlines its own cumulative-duration math. Today they're identical; the moment the helper is tightened, only the recast side follows — silent parser drift is exactly the class of bug the `#2673` parity harness exists to catch.
2. **Even-distributed arrays lost precision** (`gsapObjectArrayTiming.ts:29`). The helper rounds to integer percentages; the historical call sites emitted one-decimal (`33.3%/66.7%` for 4-element even). GSAP interprets keys literally → interior keyframes shift ~0.33% at runtime after any mutation trigger. The PR title says \"preserve duration-authored keyframe timing\" but the NO-duration path is where the actual behavior changed.

Both fix in ~10 lines total; both should land before merge.

### Cross-cutting theme: silent coercion of malformed input

Four boundary cases all silent:
- **Zero-duration entry** → tied percentages → `findObjectArrayKeyframeIndex` returns wrong index
- **Negative duration** → non-monotonic cumulative percentages, silently accepted
- **Non-numeric duration** (`duration: totalMs * 0.5`) → collapses to even distribution — the exact case the PR advertises fixing
- **Sum-of-durations ≠ outer `duration`** → per-step values dropped, mismatched outer retained (not lossless)

Full boundary test suite would take ~30 lines and catches all four.

### Semantic change worth documenting

`updateArrayKeyframeByPct` moved from index-round (always-match, clamped) to tolerance-based (`PCT_TOLERANCE=2`, returns null when nothing near enough). Studio drag paths that emit fuzzy percentages now behave differently — silent no-op instead of nearest-index update. Likely intentional, worth a comment at the call site.

Also: recast array-form update flipped from REPLACE → MERGE (necessary and correct), but object-form still REPLACES — inconsistency now permanent and undocumented in the surrounding comment.

### Family-A parity gap check (cross-stack lens)

Reading all four PR bodies against the \"1:1 Figma-motion parity\" claim in Miguel's Slack ask:

- **#2672**: 4 ease families (hold, spring, wiggle, custom). Custom is single-segment cubic-Bezier ending at (1,1) only. Overshoot (Figma's canonical anticipate/settle) is NOT expressible in this grammar (Bezier X clamp is `[0,1]`; endpoint must be exactly `1,1`). **This is a parity gap.** Either extend the grammar or Studio-side gate before commit.\n- **#2673**: parser round-trip preserves authored intent — covers Figma's duplicate-time keyframe case (multiple properties at same time; Figma tracks separately). Not a gap.\n- **#2674**: editor covers Curve / Spring / Wiggle modes; no per-property easing UI. Figma DOES have per-property easing (X and Y eased differently on same keyframe). **Potential parity gap** — not fatal for MVP but worth confirming intent.\n- **#2675**: array-form duration timing preserved. Figma's timing model IS duration-based, so this closes the array-form gap. \n\n**Net**: the stack achieves \"1:1 parity for the properties it covers,\" but *does not* cover overshoot custom curves or per-property easing. Fine if that's scoped intentionally; worth confirming in the PR bodies / discussion so the parity claim doesn't over-promise.\n\n### Contract dependencies on #2672 / #2673 / #2674\n\n- Imports nothing new from #2672/#2673/#2674. Doesn't touch ease-runtime contract. Interacts with `#2673` territory (parser authored-intent) because `objectExpressionToRecord` / `valueNodeToRecord` are shared plumbing — if `#2673`'s literal-handling changes, the `typeof record.duration === \"number\"` guard here piggybacks.\n\n### 🟢 Verified\n\n- Cumulative arithmetic terminates exactly at `100` for the last entry (`cumulative === totalDuration` at the final step → `Math.round(1.0 * 100) === 100`); no off-by-one on the endpoint.\n- All recast-side consumers of the previous even-distribution math ARE routed through the helper: `parseObjectArrayKeyframes`, `convertArrayKeyframesToObjectNode`, `removeKeyframeFromScript`, `updateKeyframeInScript`, `moveKeyframeInScript`, `resizeKeyframedTweenInScript`, `removeAllKeyframesFromScript`. All acorn-writer consumers likewise. The one gap is the Acorn PARSER at `gsapParserAcorn.ts:1022` (inline flagged).\n- Per-keyframe `duration` correctly stripped from individual entries when converting to object form, and hoisted to outer `vars.duration` when the outer key is absent — both parsers.\n- Merge on array-form update preserves sibling durations byte-for-byte via recast/magic-string localized rewrite; unmatched entries are not re-emitted.\n- Tie-break on strict `<` in `findObjectArrayKeyframeIndex` produces deterministic earliest-index-wins (only surprising in the zero-duration collision case).\n- `totalDuration === 0` branch avoids the divide-by-zero cleanly.\n\nSolid tip. The precision regression + Acorn parser miss are the two structural items; boundary coverage and the semantic-change documentation are quality-of-life. Family-A parity claim would benefit from an explicit scope note on overshoot + per-property easing.\n\n— _[Review by Rames D Jusso](https://github.com/james-russo-rames-d-jusso)_\n

File: packages/parsers/src/gsapObjectArrayTiming.ts (hunk 1 above)
Comment (miguel-heygen) [reply packages/parsers/src/gsapObjectArrayTiming.ts]: Fixed. Both duration-authored and even-distributed paths preserve tenth-percent precision; the four-entry regression is locked to 0, 33.3, 66.7, 100 in the shared-helper tests and both parser paths.

File: packages/parsers/src/gsapObjectArrayTiming.ts (hunk 2 above)
Comment (miguel-heygen) [reply packages/parsers/src/gsapObjectArrayTiming.ts]: Fixed at the shared helper boundary. Zero, negative, non-finite, expression, and partially authored duration sequences are unresolved and preserve source; conversion also rejects an authored outer duration that conflicts with the per-step sum. Each case has regression coverage.

File: packages/parsers/src/gsapWriterAcorn.ts (hunk 3 above)
Comment (miguel-heygen) [reply packages/parsers/src/gsapWriterAcorn.ts:998]: Fixed by separating operation semantics explicitly. Updates preserve historical nearest-entry fallback for in-range percentages; removals retain the named 2-percent tolerance so a fuzzy delete cannot remove the wrong entry. Mid-gap and invalid-range behavior is covered.

**Comment by miguel-heygen** [comment]
Review follow-up: both parser implementations now share one timing owner; tenth-percent precision is preserved; malformed, expression, partial, and conflicting durations preserve source; updates retain nearest-entry semantics while deletes remain tolerance-safe. This resolves the precision confirmation and partial-helper concerns raised by both reviewers.

**Review by vanceingalls — COMMENTED** [review]
## Verdict: COMMENT (post-merge adversarial R2 — one new P2 for a follow-up)

_Reviewed at head `4bfbd89d633d5fd227023643db62d2a566984edb` (post-merge). R1 was DISMISSED by the force-push that folded the R1 P2 + Rames's three findings into a single commit. Verifying the fix set, then applying the five adversarial lenses on top._

### What holds up (Rames + R1 concerns resolved)

- **Tenth-percent precision restored (R1 P2 addressed).** `roundPercentage = Math.round(p * 10) / 10` at `packages/parsers/src/gsapObjectArrayTiming.ts:6`; both authored-cumulative and even-distribute paths route through it. The four-entry regression is locked at `[0, 33.3, 66.7, 100]` in `gsapObjectArrayTiming.test.ts:10-13`. Golden churn is minimal and value-gated: `gsapParser.test.ts:1560-1561` and `gsapParserAcorn.full.test.ts:885-886` change `22→21.7`, `65→65.2`, both derived from authored `[0.5, 1, 0.8]` — not a behavior change.
- **Boundary cases (Rames #2) — rejected at helper.** `[0, 1, 1]`, `[1, -1, 1]`, `[1, NaN, 1]`, `[1, expr, 1]`, `[1, undefined, 1]` all return `null` via the `.every(d > 0 && Number.isFinite)` gate. Test matrix at `gsapObjectArrayTiming.test.ts:22-30`. Callers preserve source on `null` — verified across four sites (`gsapParser.ts:2120`, `gsapParser.ts:774`, `gsapWriterAcorn.ts:1096`, `gsapParserAcorn.ts:1076`).
- **Update/remove semantics split (Rames #3) — tolerance divergence intentional.** `findObjectArrayKeyframeIndex` accepts `{ fallbackToNearest, tolerance }`. Updates pass `fallbackToNearest: true` (`gsapParser.ts:2481`, `gsapWriterAcorn.ts:996`) — always match in range. Removes pass no options → 2% tolerance (`gsapParser.ts:2317`, `gsapWriterAcorn.ts:1272`) — fuzzy delete cannot remove the wrong entry. Precedence in the return expression is correct: `(dist <= tol) || fallback ? match : null`.

### Adversarial lenses

- **Lens A — Duration arithmetic (NEW P2).** `getCompatibleObjectArrayKeyframeTiming` at `gsapObjectArrayTiming.ts:55-59` uses `Math.abs(outerDuration - timing.totalDuration) <= Number.EPSILON`. `Number.EPSILON = 2.22e-16` is too tight for FP-sum comparisons. Empirically: `[0.7, 1.4, 2.1]` sums to `4.199999999999999`; against outer `4.2` the diff is `8.88e-16`, ~4× EPSILON → rejected. Consequences: `gsapWriterAcorn.ts:1092-1096` silently no-ops the user's array→object conversion; `gsapParser.ts:2116-2120` falls through to `convertToKeyframesInScript`, a different flat-tween shape. Not test-covered — the mismatch fixture uses `1+2+1=4` vs `8` where the diff is 4, well above any threshold. Recommend a relative tolerance like `Number.EPSILON * Math.max(1, |sum|, |outer|) * (N + 1)` or a small absolute floor (e.g. `1e-9`).
- **Lens B — Priority ordering.** Zero/negative/NaN/partial rejected at helper (verified above). `[1, 2, 1]` + outer=`4` → conversion writes cumulative `[25, 75, 100]`; when outer conflicts, `convertArrayKeyframesToObjectNode` returns `null` and `resizeKeyframedTweenInScript` (`gsapParser.ts:2432-2434`) / `moveKeyframeInScript` (`gsapParser.ts:2370-2373`) early-return. That's the tested "preserve source on conflict" contract (`gsapWriter.parity.test.ts:1342-1351`). Intentional. Retiming interference: after conversion, outer duration is written from `totalDuration` (`gsapParser.ts:2121-2123`), then resize overwrites with `newDuration` (`gsapParser.ts:2451`). Net-correct.
- **Lens C — Sibling PR #2673 (`preserve authored keyframe intent`, merged).** Both parser paths share `objectExpressionToRecord`; #2673's `mergePercentageKeyframes` and #2675's timing helper touch disjoint keyframes formats (percentage-object vs object-array). No dropped-field risk between them.
- **Lens D — Backwards compat.** All golden changes trace to explicit authored `duration:` fixtures. No implicit rerun; no invented behavior. Even-distribute test `[undefined × 4]` newly locks `[0, 33.3, 66.7, 100]` — the OLD `convertArrayKeyframesToObject*` path emitted this precision; the extraction had regressed to integer before this rev restored it.
- **Lens E — Precision propagation.** Emitters use `${timing.percentages[i]}%` string templates (`gsapParser.ts:2128`, `gsapWriterAcorn.ts:1099`) — no `toFixed`/`toPrecision` truncation. Number `33.3` stringifies verbatim. Rendered output preserves helper precision.

### Observation (P3, pre-existing, out of scope)

`parseSimpleArrayKeyframes` at `gsapParser.ts:819` and `gsapParserAcorn.ts:1117` still uses `Math.round((i / (maxLen - 1)) * 100)` — integer percentages for the zipped value-array format (`keyframes: { x: [...], y: [...] }`). Pre-existing; the PR doesn't touch it. Object-array now emits tenth-precision, simple-array remains integer — worth a follow-up ticket for consistency across the two keyframe formats.

### Counts

P1: 0 · P2: 1 (Lens A — EPSILON tolerance) · P3: 1 (simple-array precision parity, pre-existing).

— Via
